mantidproject/mantid
# SANS Reduction Non-compatibility mode

**Description of work.**
The ISIS SANS gui used to have a new reduction workflow, where workspaces were kept as event workspaces for more of the reduction process. This was removed around Dec 2018 as it was discovered that the new workflow yielded incorrect results when bin masking had occured.

This pr re-introduces the new reduction workflow.

The ISIS SANS gui has a compatibility button (default is checked) which, when checked, will run the old, correct reduction. The scientists can use the compatibility to validate the new reduction.

Therefore this PR is mainly to check code, and that the new reduction vaguely works. I have also added a system test to compare the new and old reductions with a fairly high tolerance. The scientists can later confirm whether or not this is acceptable

This PR also creates an algorithm to copy mask bins over from one workspace to another - **MaskBinsFromWorkspace**. This is necessary for the new reduction to work

**To test:**
1. Interfaces -> SANS -> ISIS SANS
2. Load the attached user and batch file
3. Click the settings tab at the left of the gui
4. Change the reduction mode to `rear`, and see that compatibility mode checkbox is on by default
5. Back on the main runs tab, click `Process All`
6. Once the row has turned green, change the `output name` in the table to **Non_compatibility mode** or something similar
7. Back in the settings tab, uncheck the compatibility mode checkbox
8. Back in the runs tab, click `Process All` again
9. plot the spectrum of the two reduced workspaces, **Compatibility mode** and **Non_compatibility mode**, and set the axes to log-log. The plots should closely align, with some small discrepancies

Fixes #23973 

**Data:**
[reduction_pr.zip](https://github.com/mantidproject/mantid/files/3237714/reduction_pr.zip)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
